File: services/dsp-service/app/pipeline/preprocessing.py
"""Streaming audio preprocessing: bandpass filter -> spectral-subtraction
noise reduction -> AGC normalization.

Both stages are implemented as true streaming, stateful processors — the
bandpass filter carries its IIR state (`zi`) across calls via
`scipy.signal.sosfilt`, and the denoiser is a real overlap-add (OLA)
short-time Fourier transform pipeline that carries both its input framing
tail and its output overlap buffer across calls. Neither stage reprocesses
audio it has already emitted, which keeps CPU cost proportional to the
incoming chunk size rather than the size of the rolling analysis window.
"""

from __future__ import annotations

from collections import deque

import numpy as np
from scipy.signal import butter, sosfilt, sosfilt_zi

from app.pipeline import constants as C


class BandpassFilter:
    """4th-order Butterworth bandpass, applied causally with persistent
    filter state so chunk boundaries introduce no discontinuity."""

    def __init__(
        self,
        sample_rate: int,
        low_hz: float = C.BANDPASS_LOW_HZ,
        high_hz: float = C.BANDPASS_HIGH_HZ,
        order: int = C.BANDPASS_ORDER,
    ) -> None:
        nyquist = sample_rate / 2.0
        low = low_hz / nyquist
        high = min(high_hz, nyquist * 0.99) / nyquist
        self.sos = butter(order, [low, high], btype="band", output="sos")
        self._zi = sosfilt_zi(self.sos) * 0.0

    def process(self, x: np.ndarray) -> np.ndarray:
        if x.size == 0:
            return x
        y, self._zi = sosfilt(self.sos, x, zi=self._zi)
        return y.astype(np.float32)


class StreamingSpectralDenoiser:
    """Overlap-add spectral subtraction (Boll, 1979) with an adaptive noise
    floor estimated from the quietest recent frames, followed by a
    frame-synchronous AGC normalization stage.

    Uses a Hann analysis *and* synthesis window at 50% hop, which satisfies
    the constant-overlap-add condition for both `sum(w)` and `sum(w^2)` —
    the standard "weighted overlap-add" configuration for spectral
    modification pipelines.
    """

    def __init__(
        self,
        frame_size: int = C.STFT_FRAME_SIZE,
        hop_size: int = C.STFT_HOP_SIZE,
    ) -> None:
        self.frame_size = frame_size
        self.hop_size = hop_size
        self.window = np.hanning(frame_size).astype(np.float32)

        self._input_tail = np.zeros(0, dtype=np.float32)
        self._output_overlap = np.zeros(0, dtype=np.float32)
        self._noise_mag: np.ndarray | None = None
        self._recent_frame_db: deque[float] = deque(maxlen=50)
        self._agc_gain = 1.0

    def _update_noise_estimate(self, mag: np.ndarray, frame_db: float) -> None:
        if self._noise_mag is None:
            self._noise_mag = mag.copy()
            self._recent_frame_db.append(frame_db)
            return

        if len(self._recent_frame_db) >= 5:
            quiet_threshold = float(np.percentile(self._recent_frame_db, C.NOISE_UPDATE_PERCENTILE))
        else:
            quiet_threshold = frame_db  # not enough history yet — treat this frame as the reference

        if frame_db <= quiet_threshold:
            alpha = C.NOISE_ESTIMATE_SMOOTHING
        else:
            alpha = 0.995  # slow leak so the estimate can still drift during long speech-only stretches

        self._noise_mag = alpha * self._noise_mag + (1 - alpha) * mag
        self._recent_frame_db.append(frame_db)

    def process(self, chunk: np.ndarray) -> np.ndarray:
        if chunk.size:
            self._input_tail = np.concatenate([self._input_tail, chunk.astype(np.float32)])

        ready_chunks: list[np.ndarray] = []

        while len(self._input_tail) >= self.frame_size:
            frame = self._input_tail[: self.frame_size]
            self._input_tail = self._input_tail[self.hop_size :]

            windowed = frame * self.window
            spectrum = np.fft.rfft(windowed)
            mag = np.abs(spectrum)
            phase = np.angle(spectrum)

            frame_db = 20.0 * np.log10(float(np.sqrt(np.mean(frame**2))) + C.EPS)
            self._update_noise_estimate(mag, frame_db)

            sub_mag = mag - C.NOISE_OVERSUBTRACTION_FACTOR * self._noise_mag
            floor = C.NOISE_SPECTRAL_FLOOR * mag
            sub_mag = np.maximum(sub_mag, floor)

            recon_spectrum = sub_mag * np.exp(1j * phase)
            recon_frame = np.fft.irfft(recon_spectrum, n=self.frame_size).astype(np.float32)

            frame_rms = float(np.sqrt(np.mean(recon_frame**2))) + C.EPS
            target_gain = float(np.clip(C.AGC_TARGET_RMS / frame_rms, C.AGC_MIN_GAIN, C.AGC_MAX_GAIN))
            self._agc_gain = C.AGC_SMOOTHING * self._agc_gain + (1 - C.AGC_SMOOTHING) * target_gain
            recon_frame = recon_frame * self._agc_gain

            synthesized = recon_frame * self.window

            out_len = max(len(self._output_overlap), self.frame_size)
            buf = np.zeros(out_len, dtype=np.float32)
            buf[: len(self._output_overlap)] = self._output_overlap
            buf[: self.frame_size] += synthesized

            ready_chunks.append(buf[: self.hop_size].copy())
            self._output_overlap = buf[self.hop_size :]

        if ready_chunks:
            return np.concatenate(ready_chunks)
        return np.zeros(0, dtype=np.float32)


class AudioPreprocessor:
    """Combines the bandpass filter and denoiser into the single per-chunk
    call site the session pipeline uses."""

    def __init__(self, sample_rate: int) -> None:
        self.bandpass = BandpassFilter(sample_rate)
        self.denoiser = StreamingSpectralDenoiser()

    def process(self, chunk: np.ndarray) -> np.ndarray:
        filtered = self.bandpass.process(chunk)
        return self.denoiser.process(filtered)


def pcm16_bytes_to_float32(raw: bytes) -> np.ndarray:
    """Decodes little-endian PCM16 mono bytes into float32 samples in [-1, 1]."""
    if len(raw) % 2 != 0:
        raw = raw[:-1]  # drop a stray trailing byte rather than fail the frame
    ints = np.frombuffer(raw, dtype="<i2")
    return (ints.astype(np.float32)) / 32768.0


def normalize_peak(x: np.ndarray, target_peak: float = 0.9) -> np.ndarray:
    """Simple one-shot peak normalization, used by the non-streaming
    calibration analysis path (see pipeline/session_pipeline.py:analyze_reference_sample)."""
    peak = float(np.max(np.abs(x))) if x.size else 0.0
    if peak < C.EPS:
        return x
    return (x * (target_peak / peak)).astype(np.float32)

File: apps/phone/src/dsp/preprocessing.ts
// Streaming audio preprocessing: bandpass filter -> spectral-subtraction
// noise reduction -> AGC normalization. Ported from
// services/dsp-service/app/pipeline/preprocessing.py. Unlike the Python
// service (which received PCM16 bytes over a WebSocket), this runs
// in-browser directly on Float32 samples already resampled to
// `settings.sampleRate` by the capture pipeline (see ../lib/pcm.ts) — so
// there is no PCM16 <-> float conversion layer here at all.

import * as C from './constants';
import { irfftFromMagPhase, rfftMagPhase } from './fft';
import { BiquadCascade, designBandpassButterworth } from './butterworth';

export class BandpassFilter {
  private readonly cascade: BiquadCascade;

  constructor(sampleRate: number, lowHz = C.BANDPASS_LOW_HZ, highHz = C.BANDPASS_HIGH_HZ, order = C.BANDPASS_ORDER) {
    const sos = designBandpassButterworth(order, lowHz, highHz, sampleRate);
    this.cascade = new BiquadCascade(sos);
  }

  process(x: Float32Array): Float32Array {
    if (x.length === 0) return x;
    return this.cascade.process(x);
  }
}

function hannWindow(n: number): Float32Array {
  const w = new Float32Array(n);
  for (let i = 0; i < n; i++) w[i] = 0.5 - 0.5 * Math.cos((2 * Math.PI * i) / n);
  return w;
}

function percentile(values: number[], p: number): number {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const idx = (p / 100) * (sorted.length - 1);
  const lo = Math.floor(idx);
  const hi = Math.ceil(idx);
  if (lo === hi) return sorted[lo];
  return sorted[lo] + (sorted[hi] - sorted[lo]) * (idx - lo);
}

function rms(frame: ArrayLike<number>): number {
  let sum = 0;
  for (let i = 0; i < frame.length; i++) sum += frame[i] * frame[i];
  return Math.sqrt(sum / frame.length);
}

/**
 * Overlap-add spectral subtraction (Boll, 1979) with an adaptive noise
 * floor estimated from the quietest recent frames, followed by
 * frame-synchronous AGC normalization — a direct port of
 * StreamingSpectralDenoiser (weighted OLA, Hann analysis+synthesis window
 * at 50% hop).
 */
export class StreamingSpectralDenoiser {
  private readonly frameSize: number;
  private readonly hopSize: number;
  private readonly window: Float32Array;

  private inputTail = new Float32Array(0);
  private outputOverlap = new Float32Array(0);
  private noiseMag: Float64Array | null = null;
  private readonly recentFrameDb: number[] = [];
  private agcGain = 1.0;

  constructor(frameSize = C.STFT_FRAME_SIZE, hopSize = C.STFT_HOP_SIZE) {
    this.frameSize = frameSize;
    this.hopSize = hopSize;
    this.window = hannWindow(frameSize);
  }

  private updateNoiseEstimate(mag: Float64Array, frameDb: number): void {
    if (this.noiseMag === null) {
      this.noiseMag = mag.slice();
      this.recentFrameDb.push(frameDb);
      if (this.recentFrameDb.length > 50) this.recentFrameDb.shift();
      return;
    }

    const quietThreshold = this.recentFrameDb.length >= 5 ? percentile(this.recentFrameDb, C.NOISE_UPDATE_PERCENTILE) : frameDb;
    const alpha = frameDb <= quietThreshold ? C.NOISE_ESTIMATE_SMOOTHING : 0.995;

    for (let i = 0; i < this.noiseMag.length; i++) {
      this.noiseMag[i] = alpha * this.noiseMag[i] + (1 - alpha) * mag[i];
    }
    this.recentFrameDb.push(frameDb);
    if (this.recentFrameDb.length > 50) this.recentFrameDb.shift();
  }

  /**
   * Returns the denoised+AGC'd audio (as before) plus one pre-AGC loudness
   * reading (dB) per emitted hop, in order -- `preAgcLoudnessDb.length *
   * hopSize === audio.length` always holds, so a caller can timestamp each
   * entry against its own running output-sample cursor (see
   * AudioPreprocessor.process()).
   *
   * The pre-AGC reading is taken from `reconFrame` (after spectral
   * subtraction, before the AGC gain stage) rather than the final AGC'd
   * output: AGC deliberately drives RMS toward a fixed AGC_TARGET_RMS, which
   * erases genuine loud-vs-quiet variation -- a signal that's been
   * normalized to a constant target can't be used to detect "the patient is
   * speaking too loud" (that's what was collapsing loudnessDb toward N/A /
   * always-or-never-triggering rather than a real value).
   */
  process(chunk: Float32Array): { audio: Float32Array; preAgcLoudnessDb: number[] } {
    if (chunk.length) {
      const merged = new Float32Array(this.inputTail.length + chunk.length);
      merged.set(this.inputTail);
      merged.set(chunk, this.inputTail.length);
      this.inputTail = merged;
    }

    const readyChunks: Float32Array[] = [];
    const preAgcLoudnessDb: number[] = [];

    while (this.inputTail.length >= this.frameSize) {
      const frame = this.inputTail.subarray(0, this.frameSize);
      this.inputTail = this.inputTail.subarray(this.hopSize);

      const windowed = new Float64Array(this.frameSize);
      for (let i = 0; i < this.frameSize; i++) windowed[i] = frame[i] * this.window[i];

      const { mag, phase } = rfftMagPhase(windowed, this.frameSize);
      const frameDb = 20 * Math.log10(rms(frame) + C.EPS);
      this.updateNoiseEstimate(mag, frameDb);

      const noiseMag = this.noiseMag!;
      const subMag = new Float64Array(mag.length);
      for (let i = 0; i < mag.length; i++) {
        const floor = C.NOISE_SPECTRAL_FLOOR * mag[i];
        subMag[i] = Math.max(mag[i] - C.NOISE_OVERSUBTRACTION_FACTOR * noiseMag[i], floor);
      }

      const reconFrame = irfftFromMagPhase(subMag, phase, this.frameSize);

      const frameRms = rms(reconFrame) + C.EPS;
      preAgcLoudnessDb.push(20 * Math.log10(frameRms));

      const targetGain = Math.min(C.AGC_MAX_GAIN, Math.max(C.AGC_MIN_GAIN, C.AGC_TARGET_RMS / frameRms));
      this.agcGain = C.AGC_SMOOTHING * this.agcGain + (1 - C.AGC_SMOOTHING) * targetGain;

      const synthesized = new Float32Array(this.frameSize);
      for (let i = 0; i < this.frameSize; i++) synthesized[i] = reconFrame[i] * this.agcGain * this.window[i];

      const outLen = Math.max(this.outputOverlap.length, this.frameSize);
      const buf = new Float32Array(outLen);
      buf.set(this.outputOverlap);
      for (let i = 0; i < this.frameSize; i++) buf[i] += synthesized[i];

      readyChunks.push(buf.slice(0, this.hopSize));
      this.outputOverlap = buf.slice(this.hopSize);
    }

    if (readyChunks.length === 0) return { audio: new Float32Array(0), preAgcLoudnessDb: [] };
    const totalLen = readyChunks.reduce((sum, c) => sum + c.length, 0);
    const out = new Float32Array(totalLen);
    let offset = 0;
    for (const chunk of readyChunks) {
      out.set(chunk, offset);
      offset += chunk.length;
    }
    return { audio: out, preAgcLoudnessDb };
  }
}

export interface LoudnessSample {
  /** Seconds, same absolute timeline as VoiceActivityDetector's FrameDecision.time (both are elapsed denoised-output samples / sampleRate, counted from session start). */
  time: number;
  /** Pre-AGC dB (see StreamingSpectralDenoiser.process()'s doc comment). */
  db: number;
}

export class AudioPreprocessor {
  private readonly bandpass: BandpassFilter;
  private readonly denoiser: StreamingSpectralDenoiser;
  private readonly sampleRate: number;
  private readonly hopSize: number;
  private outputSamplesEmitted = 0;

  constructor(sampleRate: number, hopSize: number = C.STFT_HOP_SIZE) {
    this.bandpass = new BandpassFilter(sampleRate);
    this.denoiser = new StreamingSpectralDenoiser();
    this.sampleRate = sampleRate;
    this.hopSize = hopSize;
  }

  process(chunk: Float32Array): { audio: Float32Array; loudnessSamples: LoudnessSample[] } {
    const filtered = this.bandpass.process(chunk);
    const { audio, preAgcLoudnessDb } = this.denoiser.process(filtered);

    const loudnessSamples: LoudnessSample[] = preAgcLoudnessDb.map((db, i) => ({
      time: (this.outputSamplesEmitted + i * this.hopSize) / this.sampleRate,
      db,
    }));
    this.outputSamplesEmitted += audio.length;

    return { audio, loudnessSamples };
  }
}

/** One-shot peak normalization — used by the non-streaming calibration analysis path. */
export function normalizePeak(x: Float32Array, targetPeak = 0.9): Float32Array {
  let peak = 0;
  for (let i = 0; i < x.length; i++) peak = Math.max(peak, Math.abs(x[i]));
  if (peak < C.EPS) return x;
  const scale = targetPeak / peak;
  const out = new Float32Array(x.length);
  for (let i = 0; i < x.length; i++) out[i] = x[i] * scale;
  return out;
}
